### PR TITLE
Feature/better bullets

### PR DESCRIPTION
**Pull request by SondreHus**
TLDR: Things are a but rough around the edges, but this branch is quickly becoming too big for a single feature.

- Reworked the system so it no longer initializes projectiles as gameobjects
- Made new vfx systems that replace the old gameobjects visually
- Made nice pretty muzzle flash and smoke trail 

Stuff that needs more work:
- Mesh projectile does not appear consistently when initialized, needs a bugfix
- Max-sizes of particles trips and particle numbers currently manually set in vfx instances, should be set using script for stable development
- Need a texture2D Array equivalent to allow more than one bullet to be initialized at once for future shotguns.
- Reloading is broken, but the reload system needs further implementation anyways, so not really an issue

PS: If any of you bastards are nitpicky about this pull request I will annihilate you. It is somewhat unfinished, since it is way too large as it is, and new features need to work from the new projectile baseline sooner rather than later.

PPS: I know how to make pipe bombs, don't test me.

### CODE REVIEW

**Review by toberge — CHANGES_REQUESTED** [review]
Playtested and I have some :warning: **major concerns** :warning: 

- Projectiles **don't actually damage players** anymore
- Reload doesn't work, after you've fired all your bullets you're simply **unable to fire again**
- Guns with multiple outputs are treated as if they have *one* output

While the smoke trails and muzzle flashes look great, I can't accept merging code that *actually breaks the game*. We should fix these issues promptly, though keep in mind that you don't need to be the only one working on this :sweat_smile: 

I've tried to refrain from nitpicking, but in the end I left a couple of the more serious nitpicks for you :kissing_heart:

File: Assets/Scripts/Augment/GunFactory.cs [hunk 1]
```diff
@@ -81,13 +81,31 @@ public void InitializeGun()
...
+            outputs.AddRange(extension.outputs);
+            outputs.AddRange(extension.AttachToTransforms(gunBarrel.attachmentPoints));
+            gunController.outputs = outputs.ToArray();
+        }
+        else
+        {
+            gunController.outputs = gunBarrel.outputs;
+        }
+        // TODO: The output system needs rework 
+        gunController.projectile.projectileOutput = gunController.outputs[0];
```
[the remark this answers is not in the document]
Comment (SondreHus) [reply Assets/Scripts/Augment/GunFactory.cs:105]: Yup, temporarily

File: Assets/Scripts/Augment/GunController.cs [hunk 2]
```diff
@@ -54,25 +56,7 @@ private void FireGun()
         if (stats.Ammo <= 0)
             return;
         stats.Ammo--;
-        onFire?.Invoke(stats);
```
[the remark this answers is not in the document]
Comment (SondreHus) [reply Assets/Scripts/Augment/GunController.cs:57]: Fixed

File: Assets/Scripts/Augment/BulletController.cs [hunk 3]
```diff
@@ -2,67 +2,133 @@
...
+                trailPosTexture.setValue(sampleNum * vfxPositionsPerSample + j, projectile.position);
+                UpdateProjectileMovement?.Invoke(maxDistance / (collisionSamples * vfxPositionsPerSample), ref projectile);
+            }
+
+            Collider[] collisions = ProjectileMotions.GetPathCollisions(projectile, collisionLayers);
+
+            if(collisions.Length > 0)
+            {
+                OnColliderHit?.Invoke(collisions[0], ref projectile);
+                HitboxController hitbox = collisions[0].GetComponent<HitboxController>();
```
[the remark this answers is not in the document]
Comment (SondreHus) [reply Assets/Scripts/Augment/BulletController.cs:89]: It is hitscan, it does the entirety of the path at once, one discrete step at a time, updating the vfx as it goes, cool huh?

File: Assets/Scripts/Augment/BulletController.cs [hunk 4]
```diff
@@ -2,67 +2,133 @@
...
     }
-
-    public override void Start()
+    private void Start()
     {
-        base.Start();
-        UpdateProjectileMovement += MoveWithGravity;
+        flash.transform.position = projectileOutput.position;
     }
+    public override void InitializeProjectile(GunStats stats)
```
[the remark this answers is not in the document]
Comment (SondreHus) [reply Assets/Scripts/Augment/BulletController.cs:46]: We need an entire new system to deal with this issue.

File: Assets/Scripts/Augment/GunModifier.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
+using UnityEngine;
+
+public class GunModifier : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (SondreHus) [reply Assets/Scripts/Augment/GunModifier.cs:3]: Correct, will remove it

File: Assets/Scripts/Augment/GunController.cs [hunk 6]
```diff
@@ -15,8 +15,11 @@ public class GunController : MonoBehaviour
     [HideInInspector]
     public FireRateController fireRateController;
 
+    [HideInInspector]
     public PlayerManager player;
 
+    public Transform projectileOutput;
```
[the remark this answers is not in the document]
Comment (SondreHus) [reply Assets/Scripts/Augment/GunController.cs:21]: Like mentioned, the output system needs a rework, not gonna to it in this branch

File: Assets/Scripts/Augment/GunModifier.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (toberge) [reply Assets/Scripts/Augment/GunModifier.cs:3]: Um sorry but it's still here?

**Review by Fueredoriku — CHANGES_REQUESTED** [review]
I have encountered som unmentioned issues:
- Guns that shoot meshes (hat) seem to have the particle it clones active and viewable when initialized:
image
- The gun always initializes with two extensions instead of 1, shown below with the fire extension:
image
- Fire extension doesn't instantiate flames onHit anymore? 🤔

**Review by toberge — APPROVED** [review]
Needs a few adjustments that could be done later (and should be added as issues), but it's good enough to replace our current _very bad_ projectiles :))))))

**Review by Fueredoriku — APPROVED** [review]
Seems like the double extension problem has been resolved! 👍

We'll make the fire extension update it's own separate issue.

Tested and works on my machine, great work! 🤠

**Review by SondreHus — COMMENTED** [review]
Fixed
